src/main/scala/org/virtuslab/internal/load/compose/Composer.scala

```diff
@@ -0,0 +1,98 @@
+package org.virtuslab.internal.load.compose
+
+import org.virtuslab.internal.YamlError
+import org.virtuslab.internal.load.{YamlReader, StringYamlReader}
+import org.virtuslab.internal.load.parse.{Event, ParserImpl}
+import org.virtuslab.internal.load.compose.Node
+
+import scala.annotation.tailrec
+
+trait Composer:
+  def compose(yaml: String): Either[YamlError, Node]
+  def compose(reader: YamlReader): Either[YamlError, Node]
+
+object ComposerImpl extends Composer with NodeTransform:
+
+  type ComposeResult = (Either[YamlError, Node], List[Event])
+
+  override def compose(yaml: String): Either[YamlError, Node] = compose(StringYamlReader(yaml))
+
+  override def compose(reader: YamlReader): Either[YamlError, Node] =
+    for
+      events <- ParserImpl.getEvents(reader, ???)
+      node   <- fromEvents(events)
+    yield node
+
+  override def fromEvents(events: List[Event]): Either[YamlError, Node] =
+    events match
+      case Nil =>
+        Left(YamlError("No events available"))
+      case _ =>
+        val (node, _) = composeNode(events)
+        node
+
+  private def composeNode(events: List[Event]): ComposeResult =
+    events match
+      case head :: tail =>
+        head match
+          case Event.StreamStart | Event.DocumentStart => composeNode(tail)
+          case Event.SequenceStart               => composeSequenceNode(tail)
+          case Event.MappingStart                => composeMappingNode(tail)
+          case s: Event.Scalar                   => (composeScalarNode(s), tail)
+          case event                       => (Left(YamlError(s"Unexpected event $event")), events)
+      case Nil => (Left(YamlError("No events available")), Nil)
+
+  private def composeSequenceNode(events: List[Event]): ComposeResult = {
+    @tailrec
+    def parseChildren(
+        events: List[Event],
+        children: List[Node]
+    ): (Either[YamlError, List[Node]], List[Event]) = {
+      events match
+        case Nil                 => (Left(YamlError("No events available")), Nil)
+        case Event.SequenceEnd :: tail => (Right(children), tail)
+        case _ =>
+          val (node, rest) = composeNode(events)
+          node match
+            case Right(value) => parseChildren(rest, children :+ value)
+            case Left(err)    => (Left(err), rest)
```

Comment:
In any case of error we should skip any further parsing, so why is `rest` returned here?

Reply:
Yeah, that could be `Nil`. It' not ideal that the result is Tuple[Either, List] because for Left there is no sense in returning a List of remaining events but this could be improved later